### PR TITLE
Ext 612 assign templates to projects

### PR DESCRIPTION
**Pull request by genevieveluyt**

### CODE REVIEW

File: src/js/popup.js [hunk 1]
```diff
@@ -21,28 +21,44 @@ if (navigator.userAgent.indexOf('Firefox') !== -1 || navigator.userAgent.indexOf
...
+        } else {
+            return 0;
         }
-    }
-    a.sort();
+    });
+}
+
+function sortTemplates (templates) {
+    var templateArray = $.map(templates, function (template, key) {
```
Comment (genevieveluyt) [inline src/js/popup.js]: templates is an object

File: src/js/popup.js [hunk 2]
```diff
@@ -73,41 +88,80 @@ function loadTemplateEditor (callback = false) {
...
-                    templateTitle = '<div class="collapsible-header grey lighten-5" data-issueFieldType="' + template['issuetype-field'] + '" style="font-weight: bold;"><i class="material-icons">expand_less</i>' + key + '</div>';
-                } else {
-                    templateTitle = '<div class="collapsible-header grey lighten-5" data-issueFieldType="' + template['issuetype-field'] + '"><i class="material-icons">expand_less</i>' + key + '</div>';
+            $.each(templatesArray, function (index, template) {
+                var boldStyle = '';
+                if (!template['issuetype-field'] && !template['projects-field']) {
+                    boldStyle = 'style="font-weight: bold;"';
                 }
 
+                var templateTitle =
```
Comment (genevieveluyt) [inline src/js/popup.js]: Ooh that definitely looks a lot nicer! As far as I can tell it is just vanilla JS so I'll make that change.

File: src/jti_background.js [hunk 3]
```diff
@@ -140,63 +143,104 @@ function setDefaultTemplates (callback) {
...
+    });
+}
 
-        if (save) {
-            templateJSON.templates[templateName] = {'issuetype-field': issueTypeField, 'text': text};
-            saveTemplates(templateJSON, callback);
+// Make sure that the name and (issue type, project) combination is unique
+function validateTemplate (newTemplate, templates, callback) {
+    var valid = true;
+    $.each(templates, function (name, template) {
```
Comment (genevieveluyt) [inline src/jti_background.js]: It's not, it's just a placeholder. Is there a better way to do that?

File: src/js/popup.js [hunk 4]
```diff
@@ -116,13 +170,17 @@ function loadTemplateEditor (callback = false) {
...
+                var bgPage = chrome.extension.getBackgroundPage();
                 var defaultTemplates = response.data.templates;
 
-                // Remove default templates from dropdown list if already added.
-                $.each(dropdownExcludeList, function (index, issueTypeField) {
-                    $.each(defaultTemplates, function (name, template) {
-                        if (issueTypeField === template['issuetype-field']) {
-                            delete defaultTemplates[name];
+                // Remove default templates from dropdown list if a template for that (issue type, projects) combination already exists.
+                $.each(templatesArray, function (index, excludeTemplate) {
```
Comment (genevieveluyt) [inline src/js/popup.js]: I ended up using Handlebars which is specifically for templating. Lodash seems a little overkill.

File: README.md [hunk 5]
```diff
@@ -15,13 +15,15 @@ The JSON format is:
 
 ```javascript
 {
-	"templates": {
-    	"NAME_OF_TEMPLATE": {
+	"templates": [
+    	{
+            "name":"NAME_OF_TEMPLATE",
         	"issuetype-field":"issue_type_field",
```
Comment (genevieveluyt) [inline README.md]: :+1:

File: src/js/popup.js [hunk 6]
```diff
@@ -73,41 +88,80 @@ function loadTemplateEditor (callback = false) {
...
-            $.each(templates, function (key, template) {
-                var templateTitle = '';
-                if (key === 'DEFAULT TEMPLATE') {
-                    templateTitle = '<div class="collapsible-header grey lighten-5" data-issueFieldType="' + template['issuetype-field'] + '" style="font-weight: bold;"><i class="material-icons">expand_less</i>' + key + '</div>';
-                } else {
-                    templateTitle = '<div class="collapsible-header grey lighten-5" data-issueFieldType="' + template['issuetype-field'] + '"><i class="material-icons">expand_less</i>' + key + '</div>';
+            $.each(templatesArray, function (index, template) {
+                var boldStyle = '';
+                if (!template['issuetype-field'] && !template['projects-field']) {
+                    boldStyle = 'style="font-weight: bold;"';
```
Comment (genevieveluyt) [inline src/js/popup.js]: :+1:

File: src/js/popup.js [hunk 7]
```diff
@@ -116,13 +170,17 @@ function loadTemplateEditor (callback = false) {
         // Populate the add default template dropdown - excluding any templates already loaded.
         chrome.runtime.sendMessage({JDTIfunction: 'fetchDefault'}, function (response) {
             if (response.status === 'success') {
+                var bgPage = chrome.extension.getBackgroundPage();
```
Comment (genevieveluyt) [inline src/js/popup.js]: :+1:

File: src/js/popup.js [hunk 8]
```diff
@@ -553,7 +610,7 @@ $(document).ready(function () {
             dmUIClick('remove-single');
             chrome.runtime.sendMessage({
                 JDTIfunction: 'delete',
-                templateName: $(this).attr('id')
+                templateName: $(this).attr('template')
```
Comment (genevieveluyt) [inline src/js/popup.js]: :+1:

File: src/jti_background.js [hunk 9]
```diff
@@ -140,63 +143,104 @@ function setDefaultTemplates (callback) {
...
+            templateJSON.templates[templateID] = newTemplate;
+            saveTemplates(templateJSON, callback, templateID);
+        }
+    });
+}
 
-        if (save) {
-            templateJSON.templates[templateName] = {'issuetype-field': issueTypeField, 'text': text};
-            saveTemplates(templateJSON, callback);
+// Make sure that the name and (issue type, project) combination is unique
```
Comment (genevieveluyt) [inline src/jti_background.js]: :+1:

File: .eslintrc.js [hunk 10]
```diff
@@ -19,5 +19,8 @@ module.exports = {
         "indent": ["error", 4],
         "semi": ["error", "always", { "omitLastInOneLineBlock": true}],
         "one-var": "off"
+    },
+    "globals": {
+        "utils": false
     }
 };
```
Comment (cfoucher) [inline .eslintrc.js]: new line

File: README.md [hunk 11]
```diff
@@ -39,8 +41,9 @@ The JSON format is:
 * Automatically select text between ```<TI>``` and ```</TI>``` tags for quick template completion.
   * Quickly jump to the next set of ```<TI>``` elements using the ```Control + back-tick``` key combo.
 * Template Priorities.
-  * Set a default template (DEFAULT TEMPLATE) for all Issue Types.
-  * Individual Issue Types will override the default.
+  * Set a default template (DEFAULT TEMPLATE) for all Issue Types and Projects.
+  * Templates with Issue Type will override the default.
+  * Templates with Issue Type and Projects will override templates with just Issue Type.
```
Comment (cfoucher) [inline README.md]: What about Templates with a Project but no Issue Type?

File: src/css/popup.css [hunk 12]
```diff
@@ -307,4 +311,8 @@ textarea.materialize-textarea:focus:not([readonly]) {
 
 .btn-floating i {
 	font-size: 1.3rem;
+}
+
+.font-bold {
+	font-weight: bold;
 }
```
Comment (cfoucher) [inline src/css/popup.css]: New line

File: src/html/popup.html [hunk 13]
```diff
@@ -209,5 +219,8 @@ <h5 id="addTemplateTitle">Add Default or Custom Template</h5>
                 <!-- Dynamically built -->
             </ul>
         </main>
+
+        <!-- Need this to be able to grab the sandbox window to compile the edit templates -->
+        <iframe id="sandbox_window" src="sandbox.html" hidden></iframe>
     </body>
 </html>
```
Comment (cfoucher) [inline src/html/popup.html]: New line

File: src/html/sandbox.html [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+                    event.source.postMessage({
+                        html: htmlTemplate(event.data.context),
+                        callback: event.data.callback,
+                        openTemplate: event.data.openTemplate
+                    }, event.origin);
+                }
+            });
+        </script>
+    </body>
+</html>
```
Comment (cfoucher) [inline src/html/sandbox.html]: New line.

File: src/html/popup.html [hunk 15]
```diff
@@ -209,5 +219,8 @@ <h5 id="addTemplateTitle">Add Default or Custom Template</h5>
                 <!-- Dynamically built -->
             </ul>
         </main>
+
+        <!-- Need this to be able to grab the sandbox window to compile the edit templates -->
+        <iframe id="sandbox_window" src="sandbox.html" hidden></iframe>
```
Comment (cfoucher) [inline src/html/popup.html]: Why use an iframe/sandbox?

Comment (cfoucher) [inline src/html/popup.html]: Never mind. Did some research on it, and see why. 
👍

**Pull request merged**